src/main/java/com/zendesk/maxwell/producer/MaxwellPubsubProducer.java

```diff
@@ -0,0 +1,212 @@
+
+    cc.markCompleted();
+    publishMessageTimer.update(cc.timeSinceSendMS(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void onFailure(Throwable t) {
+    this.failedMessageCount.inc();
+    this.failedMessageMeter.mark();
+
+    LOGGER.error(t.getClass().getSimpleName() + " @ " + position);
+    LOGGER.error(t.getLocalizedMessage());
+
+    if ( !this.context.getConfig().ignoreProducerError ) {
+      this.context.terminate(new RuntimeException(t));
+      return;
+    }
+
+    cc.markCompleted();
+    publishMessageTimer.update(cc.timeSinceSendMS(), TimeUnit.MILLISECONDS);
+  }
+
+}
+
+
+public class MaxwellPubsubProducer extends AbstractProducer {
+  public static final Logger LOGGER = LoggerFactory.getLogger(MaxwellPubsubProducer.class);
+
+  private final ArrayBlockingQueue<RowMap> queue;
+  private final MaxwellPubsubProducerWorker worker;
+
+  public MaxwellPubsubProducer(MaxwellContext context, Properties pubsubProperties, String pubsubProjectId, String pubsubTopic) throws IOException {
+    super(context);
+    this.queue = new ArrayBlockingQueue<>(100);
+    this.worker = new MaxwellPubsubProducerWorker(context, pubsubProperties, pubsubProjectId, pubsubTopic, this.queue);
+    Thread thread = new Thread(this.worker, "maxwell-pubsub-worker");
+    thread.setDaemon(true);
+    thread.start();
+  }
+
+  @Override
+  public void push(RowMap r) throws Exception {
+    this.queue.put(r);
+  }
+
+  @Override
+  public StoppableTask getStoppableTask() {
+    return this.worker;
+  }
+}
+
+
+class MaxwellPubsubProducerWorker extends AbstractAsyncProducer implements Runnable, StoppableTask {
+  static final Logger LOGGER = LoggerFactory.getLogger(MaxwellPubsubProducer.class);
+
+  private final String projectId;
+  private Publisher pubsub;
+  private final TopicName topic;
+  private final TopicName ddlTopic;
+  private Publisher ddlPubsub;
+  private final ArrayBlockingQueue<RowMap> queue;
+  private Thread thread;
+  private StoppableTaskState taskState;
+
+  public MaxwellPubsubProducerWorker(MaxwellContext context, Properties pubsubProperties, String pubsubProjectId, String pubsubTopic, ArrayBlockingQueue<RowMap> queue) throws IOException {
+    super(context);
+
+    if ( pubsubTopic == null ) {
+      this.topic = TopicName.create(pubsubProjectId, "maxwell");
+    } else {
+      this.topic = TopicName.create(pubsubProjectId, pubsubTopic);
+    }
+
+    this.projectId = pubsubProjectId;
+    this.ddlTopic = TopicName.create(pubsubProjectId, context.getConfig().ddlPubsubTopic);
+
+    this.setPubsub();
+    this.setDDLPubsub();
+
+    Metrics metrics = context.getMetrics();
+
+    this.queue = queue;
+    this.taskState = new StoppableTaskState("MaxwellPubsubProducerWorker");
+  }
+
+  public void setPubsub() throws IOException {
+    this.pubsub = Publisher.defaultBuilder(this.topic).build();
+  }
+
+  public void setDDLPubsub() throws IOException {
+    this.ddlPubsub = Publisher.defaultBuilder(this.ddlTopic).build();
+  }
```

Comment:
Can these be private / inlined now? (I'm not sure I understand the intent here, as they don't allow any control over what instance gets created)

Reply:
Also, if the creation of a publisher causes some external action (e.g. connecting to a broker or validating the topic name exists), we shouldn't set one up unless `context.getConfig().outputDDL == true`

Reply:
Welp! Yep, updated.